cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java

```diff
@@ -143,8 +144,15 @@ public KafkaConnectAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pf
                 .compose(i -> deploymentOperations.reconcile(namespace, connect.getName(), connect.generateDeployment(annotations, pfa.isOpenshift(), imagePullPolicy, imagePullSecrets)))
                 .compose(i -> deploymentOperations.scaleUp(namespace, connect.getName(), connect.getReplicas()))
                 .compose(i -> deploymentOperations.waitForObserved(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> deploymentOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs))
-                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus))
+                .compose(i -> connect.getReplicas() > 0 ? deploymentOperations.readiness(namespace, connect.getName(), 1_000, operationTimeoutMs) : Future.succeededFuture())
+                .compose(i -> reconcileConnectors(reconciliation, kafkaConnect, kafkaConnectStatus)
+                        .recover(error -> {
+                            if (error instanceof ConnectTimeoutException && connect.getReplicas() == 0)   {
```

Comment:
I might be wrong in saying this, but maybe we should tell `reconcileConnectors` to expect a `ConnectTimeoutException` if `replicas ==  0`? Then we don't have to capture the exception here.

(essentially move this logic into reconcileConnectors and change its behaviour.)

Reply:
I had it that way originally. But it is not completely easy, because the connector reconciliation is triggered form two sources:
* `KafkaConnectAssemblyOperator` and `KafkaConnectS2IAssemblyOperator` which can provide the replicas.
* The KafkaConnector watch which doesn't have them. So it creates inconsistent results based on where it is triggered from and it seemed to be better this way.

Reply:
Won't this error handling result in the `Ready` status on the `KafkaConnect`? It's ready in the sense that there wasn't an error in the reconciliation, but not in the sense that the cluster is 'ready' for REST requests, is that what we want?

Reply:
It does, but that is the point of the whole issue. To stop the reocnciliation failing when scaled to 0. There will be no REST API, but the resource will be ready in the sense that it will run as requested from the user.

Reply:
Fair point. But that means we're treating `NotReady` has having the semantics of "failed", rather than simply the cluster not being ready. Maybe we need a different condition entirely? I do feel that we need some way to distinguish a scaled-to-zero-like-you-asked kind of Ready from REST-interface-should-work kind of Ready, because people will want to automate on top of this status and so it needs to be unambiguous.

Reply:
So what status would you set in such case?

Reply:
I'm not sure, are there any precedents in Kube with scale to zero (in `Deployment`, `ReplicaSet` or `StatefulSet`)? If Kube uses `Ready` in this case, maybe it would be enough to omit the REST connection information when scaled to zero.

Reply:
Only Pod has `Ready` conditionm but there is no pod when it has 0 replicas. In deployment, the conditions do not show anything different from when they have replicas. They have for example the `Available` condition set to `True`.

Reply:
I think it would make sense to me that we use a similar mechanism to what deployments and statefulsets use and say `Available` false and give a reason like: `NoReplicas` / `InvalidReplicas`, e.g. the condition would look something like:
```  
 - lastTransitionTime: "2020-04-29T09:26:39Z"
    lastUpdateTime: "2020-04-29T09:26:39Z"
    message: Scaling KafkaConnect to 0 is not supported.
    reason: InvalidReplicas
    status: "False"
    type: Available

```

Reply:
They actually don't seem to do that. A deployment scaled to 0 has just this:

```
  status:
    conditions:
    - lastTransitionTime: "2020-04-29T08:15:31Z"
      lastUpdateTime: "2020-04-29T08:34:48Z"
      message: ReplicaSet "strimzi-cluster-operator-7bcd647d8d" has successfully progressed.
      reason: NewReplicaSetAvailable
      status: "True"
      type: Progressing
    - lastTransitionTime: "2020-04-29T11:18:52Z"
      lastUpdateTime: "2020-04-29T11:18:52Z"
      message: Deployment has minimum availability.
      reason: MinimumReplicasAvailable
      status: "True"
      type: Available
    observedGeneration: 3
```

So it actually says it is available. If we would go for it (regardless what Deployment does), the wierd thing would be whether it make sense to say Available / False when we do not have available / True when everything is working.

Reply:
Thinking more I don't think we should try to solve this using conditions, because we'll be forced into have two and that makes life harder at the point of use because to test for readiness you would need to check for both `Ready` and `Available` (or Ready and not Unavailable, or whatever). replicas=0 is a bit of an edge case which most people shouldn't have to deal with. 

So on reflection when replicas=0 I guess it would be OK to say `Ready=true`, but to not set the `status.url` or `status.connectorPlugins`. I think that would be enough of a signal for people wanting to use replicas=0.

Reply:
Makes sense to me, my original thinking was that we would effectively override the deployments state of availability (that's why my suggestion differs from the deployment example Jakub linked) - but you are right querying two separate status conditions seems like extra complexity...

Tom's suggestion makes sense to me, though I do wonder if a user would be confused why the url/connectorPlugins fields are missing. Perhaps we follow the pattern that @dalelane introduced in #2663 and print an additional condition as a warning so the user has a little more to go on?

Reply:
> Tom's suggestion makes sense to me, though I do wonder if a user would be confused why the url/connectorPlugins fields are missing. Perhaps we follow the pattern that @dalelane introduced in #2663 and print an additional condition as a warning so the user has a little more to go on?

I can do that if you want. Although the scaling to 0 is from my point of view always intentional and not done by mistake or not understanding how things work.

Reply:
Very true, I agree it seems very unlikely a user would scale down to 0 and also expect a service to still exist. But i do still think an additional entry to conditions clarifying that ita ready but only in the trivial sense of that all containers (that is zero of them) are ready as there is nothing to do.

Reply:
What's the proposal exactly:

```yaml
type: "Available"
status: "False"
reason: "ScaledToZero"
```

Reply:
I'm not keen on calling it a warning. They asked for 0, and we gave them 0. There's no particular reason to treat that differently from other numbers of replicas. They asked for 3, but maybe they meant 5?

I think the warnings we added in #2663 were useful because they highlighted very definite cases where people could break their clusters/lose data. That's not the case here.

Reply:
I'm not sure I like the _Available: False_ when we have no _Available: True_. But if that is fine for @samuel-hawker I can implement it.

Reply:
I agree that we should forego the Available condition, I wish I had made that a bit more clear further up the thread.
I agree that if a user asks for 0 replicas, we give them 0 replicas, I suggested a Warning only because a user in theory could wonder why the `status.url` or `status.connectorPlugins` fields are empty or missing.

Something like:
```
type: "Warning"
reason: "Zero replicas so no provided URL"
```
But maybe there is little value in this - what do you think @tombentley ?